ANALYSIS/ANALYSIS/AliAnalysisManager.cxx

```diff
@@ -3139,3 +3148,92 @@ void AliAnalysisManager::InitInputData(AliVEvent* esdEvent, AliVfriendEvent* esd
+if ( iAddedFiles < 1 ) { std::cout << "NO Files added to collection !!!" << std::endl; return NULL; }
+
+// if cTreeNameArg is auto lets try to autodetect what type of tree we have;
+// the assuption is that all files are the same and the first one is reprezentative
+THashList* list =  filecoll.GetList();
+if ( sTreeNameArg.EqualTo("auto") ) { // if tree name is not specified
+  TRegexp tree_regex ("[aod,esd]Tree");
+  TFileInfo* fileinfo = dynamic_cast<TFileInfo*>(list->At(0)); // get first fileinfo in list
+  TFile file (fileinfo->GetFirstUrl()->GetFile()); // get the actual TFile
+  if (file.IsZombie()) { cout << "Should not reach this message!! Error opening file" << endl; return NULL; }
+
+  // lets parse the TFile
+  TIter next(file.GetListOfKeys());
+  TKey* key = NULL;
+  while (( key = dynamic_cast<TKey*>(next()) )) {
+    TString class_name = key->GetClassName();
+    if ( ! class_name.EqualTo("TTree") ) { continue; } // searching for first TTree
+
+    TString key_name = key->GetName();
+    if ( key_name.Contains(tree_regex) ) { treeName = key_name; break;} // that is named either aodTree or esdTree
+    }
+  file.Close();
+  }
+else
+  { treeName = sTreeNameArg ; } // tree name is specified
+
+TChain* chain = new TChain (treeName.Data(),""); // lets create the chain
+if ( chain->AddFileInfoList(list) == 0 ) { return NULL; } // and add file collection (THashList is a Tlist that is a TCollection)
+
+// start treatment of friends
+TChain* chainFriend = NULL;
+TString sFriends (friends);
+if (!sFriends.IsNull()) {
+  TString friends_treename, friends_filename;
+  TObjArray* arr = sFriends.Tokenize("/");
+  TObjString* strobj_file = dynamic_cast<TObjString*>(arr->At(0));
+  if (strobj_file) { friends_filename = strobj_file->GetString(); }
+  TObjString* strobj_tree = dynamic_cast<TObjString*>(arr->At(1));
+  if (strobj_tree) { friends_treename = strobj_tree->GetString(); }
+  delete arr;
+
+  if (friends_treename.IsNull()) {
+    if (treeName.EqualTo("esdTree")) {
+      friends_treename = "esdFriendTree"; }
+    else if (treeName.EqualTo("aodTree")) {
+      friends_treename = "aodTree"; }
+    else {
+      cout << "friends argument specified but tree name neither specified nor auto-detected (unknown tree name to associate with known friend tree name)";
+      return chain; // stop processing of friends, just return the chain created so far
+      }
+    }
+
+  chainFriend = new TChain(friends_treename.Data());
+  TString friendinfo_for_chain = "/" + friends_filename + "/" + friends_treename;
+
+  TIter next(list);
+  TFileInfo* fileinfo = NULL;
+  while (( fileinfo = dynamic_cast<TFileInfo*>(next()) )) {
+    TString dirname = gSystem->DirName(fileinfo->GetFirstUrl()->GetFile());
+    TString friend_for_chain = dirname + friendinfo_for_chain;
+    if (chainFriend) { chainFriend->Add(friend_for_chain.Data()); }
+    }
+
+  if (chainFriend) { chain->AddFriend(chainFriend); }
+  }
```

Comment:
How is such chainFriend will be used? Note that the AnalysisManager reads the friends file by file (whenever the esd file in the esd chain changes) and knows how to associate the tracks in the esd and esdfriends. It also knows how to reset the friend events: the TPC seeds in the friends are using shared cluster objects which need to be properly cleared before reading next event. Reading the friends via standard trees mechanism is error prone.

Reply:
i do not use friends and i have no test case for them but i tried to keep the logic of https://github.com/alisw/AliPhysics/blob/master/PWG/EMCAL/macros/CreateAODChain.C and https://github.com/alisw/AliPhysics/blob/master/PWG/EMCAL/macros/CreateESDChain.C

Reply:
N.B. if friends option is not used this code path is not used

Reply:
Also, this is just a static function used to replace the macro that creates chain for the analysis .. so, given that AliAnalysisManager needs a chain to work on (to start analysis), i do not think that does anything automatically (i do not see any StartAnalysis signature from AliAnalysisManager that use as argument a list of files)

Reply:
well, you are the boss... i actually have this function in my task so i can use it anyway .. i just thought that would be of use to larger public and this is why i did this pr.
there are 2 options :   
i can remove the treatment of friends   
or   
just close the pr and let you do it.
imho the way i do it is more efficient and does also autodetection of the tree type (and because of TFileCollection the comments and validity of files are managed automatically)

Reply:
@adriansev : I am not the boss, I just see that the friends part of the patch may lead to crash. If you want to keep the signature you used, it is ok, but please remove the friends part.
Cheers,
 Ruben